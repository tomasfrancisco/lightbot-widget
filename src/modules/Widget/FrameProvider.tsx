import { theme as globalTheme } from "@lightbot/widget-configuration";
import * as React from "react";
import { FrameContextConsumer } from "react-frame-component";
import { StyleSheetManager, ThemeProvider } from "styled-components";
import * as WebFont from "webfontloader";

const getProvider = props => ({ document, window }) => {
  WebFont.load({
    context: window,
    google: {
      families: ["Assistant"],
    },
  });

  return (
    <StyleSheetManager target={document.head}>
      <ThemeProvider theme={globalTheme}>{props.children}</ThemeProvider>
    </StyleSheetManager>
  );
};

export const FrameProvider = props => (
  <FrameContextConsumer>{getProvider(props)}</FrameContextConsumer>
);
