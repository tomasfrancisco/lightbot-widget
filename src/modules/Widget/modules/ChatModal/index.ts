export * from "./ChatModal";
export * from "./stores";
