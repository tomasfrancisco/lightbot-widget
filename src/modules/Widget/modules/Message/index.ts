export * from "./Message.module";
export * from "./Message.type";
