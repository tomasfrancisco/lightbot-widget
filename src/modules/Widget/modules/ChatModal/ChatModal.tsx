import { Chat, ChatBody, ChatFooter, ChatHeader } from "components/Chat";
import { CloseButton } from "components/CloseButton";
import { MessageGroup } from "components/Message";
import { MessageComposer } from "components/MessageComposer";
import _get from "lodash.get";
import { reaction, toJS } from "mobx";
import { inject, observer } from "mobx-react";
import * as React from "react";
import CloseIcon from "react-icons/lib/fa/chevron-down";
import { v4 as uuid } from "uuid";

import {
  MessageModule as Message,
  MessageSenderEnum,
  MessageType,
  MessageTypeEnum,
} from "../Message";
import { fetchingMessage } from "../Message/constants";
import { ChatStore } from "./stores/ChatStore";

const footerHeight = "60px";

export type ChatModelMessageGroup = {
  touched?: boolean;
  messages: MessageType[];
  sender: MessageSenderEnum;
};
export type ChatModalMessages = ChatModelMessageGroup[];

@inject("chatStore")
@observer
export class ChatModal extends React.Component<{
  chatStore?: ChatStore;
  onSubmit: (message: string, label?: string, type?: MessageTypeEnum) => void;
}> {
  private messageComposerInputRef;

  constructor(props) {
    super(props);
  }

  public componentDidMount() {
    const { chatStore } = this.props;
    if (!chatStore) {
      throw Error("ChatModal could not be initialized without a provided ChatStore");
    }

    this.watchMessageComposerChanges(chatStore);
  }

  public render() {
    const { chatStore, onSubmit } = this.props;

    if (!chatStore) {
      return null;
    }

    const { agentData$, messagesCount$, isTyping$ } = chatStore;

    return (
      <Chat>
        <ChatHeader>
          <span>{agentData$.name}</span>
          {this.renderCloseButton()}
        </ChatHeader>
        <ChatBody messagesCount={messagesCount$} footerHeight={footerHeight}>
          {this.renderMessages()}
        </ChatBody>
        <ChatFooter height={footerHeight}>
          <MessageComposer
            assignRef={this.assignMessageComposerInputRef}
            touched={isTyping$}
            onSubmit={onSubmit}
            replyPlaceholder={agentData$.widgetInputPlaceholder}
          />
        </ChatFooter>
      </Chat>
    );
  }

  private renderCloseButton() {
    const { chatStore } = this.props;
    if (!chatStore) {
      return null;
    }
    const { isMobileVersion$ } = chatStore;
    return (
      isMobileVersion$ && (
        <CloseButton onClick={this.onCloseButtonClick}>
          <CloseIcon />
        </CloseButton>
      )
    );
  }

  private renderMessages() {
    const { chatStore } = this.props;
    if (!chatStore) {
      return null;
    }

    const { messages$, isTyping$ } = chatStore;

    const messageHistory = toJS(messages$).map((messageGroup: ChatModelMessageGroup, i) => (
      <MessageGroup
        key={uuid()}
        onClick={this.getOnMessageGroupClickHandler(i)}
        {...this.getMessageUIProps(messageGroup.sender)}
      >
        {this.renderMessage(messageGroup)}
      </MessageGroup>
    ));

    const typingMessage = isTyping$ ? (
      <MessageGroup key={uuid()} {...this.getMessageUIProps(MessageSenderEnum.BOT)}>
        <Message
          key={uuid()}
          message={fetchingMessage}
          sender={MessageSenderEnum.BOT}
          onClick={this.props.onSubmit}
        />
      </MessageGroup>
    ) : null;

    return (
      <React.Fragment>
        {messageHistory}
        {typingMessage}
      </React.Fragment>
    );
  }

  private renderMessage = (messageGroup: ChatModelMessageGroup) => {
    return messageGroup.messages.map(message => (
      <Message
        key={uuid()}
        message={message}
        sender={messageGroup.sender}
        onClick={this.props.onSubmit}
        touched={messageGroup.touched}
      />
    ));
  };

  private assignMessageComposerInputRef = messageComposerInputRef => {
    this.messageComposerInputRef = messageComposerInputRef;
  };

  private watchMessageComposerChanges = (store: ChatStore) => {
    reaction(
      () => {
        return store.isOpen$ && !store.isTyping$;
      },
      () => {
        this.focusOnMessageComposer();
      },
      {
        equals: (from, to) => from && !to,
        fireImmediately: true,
      },
    );
  };

  private focusOnMessageComposer = () => {
    if (!this.messageComposerInputRef) {
      return;
    }
    setTimeout(() => this.messageComposerInputRef.focus(), 200);
  };

  private getMessageUIProps(sender: MessageSenderEnum) {
    const { chatStore } = this.props;

    const agentName = _get(chatStore, ["agentData$", "name"], "");

    switch (sender) {
      case MessageSenderEnum.HUMAN:
        return {
          borderRadius: "25px",
          isRtl: true,
        };
      case MessageSenderEnum.BOT:
        return {
          borderRadius: "25px",
          isRtl: false,
          secondary: true,
          showAvatar: true,
          avatarName: agentName,
        };
      default:
        return {};
    }
  }

  private onCloseButtonClick = () => {
    const { chatStore } = this.props;
    if (chatStore) {
      chatStore.toggle$();
    }
  };

  private getOnMessageGroupClickHandler(messageGroupIndex: number) {
    return event =>
      _get(event, ["target", "tagName"]) === "BUTTON"
        ? this.touchedMessageGroupActions(messageGroupIndex)
        : null;
  }

  private touchedMessageGroupActions = (messageGroupIndex: number) => {
    if (this.props.chatStore) {
      this.props.chatStore.touchedMessageGroup(messageGroupIndex);
    }
  };
}
