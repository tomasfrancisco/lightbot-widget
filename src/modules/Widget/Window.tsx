import { inject, observer } from "mobx-react";
import * as React from "react";
import styled from "styled-components";
import { ThemeProvider } from "../ThemeProvider/ThemeProvider";
import { ChatModal } from "./modules/ChatModal";
import { ChatStore } from "./modules/ChatModal/stores/ChatStore";
import { MessageSenderEnum, MessageTypeEnum } from "./modules/Message";

const GlobalStyles = styled.div`
  & * {
    font-family: "Assistant", sans-serif;
  }

  & p {
    margin: 0;
  }
`;

@inject("chatStore")
@observer
export class Window extends React.Component<
  {
    chatStore?: ChatStore;
    agentId: string;
  },
  {}
> {
  public render() {
    if (!this.props.chatStore) {
      return null;
    }

    const { started$, agentData$ } = this.props.chatStore;

    if (!started$) {
      return null;
    }
    return (
      <ThemeProvider theme={agentData$.widgetThemeData}>
        <GlobalStyles>
          <ChatModal onSubmit={this.onSubmitMessage} />
        </GlobalStyles>
      </ThemeProvider>
    );
  }

  private onSubmitMessage = async (message: string, label?: string, type?: MessageTypeEnum) => {
    const { chatStore } = this.props;

    if (chatStore) {
      if (type === MessageTypeEnum.JUMP) {
        if (!label) {
          throw new Error("A Jump was triggered without providing a proper label.");
        }

        chatStore.appendJumpMessage(message, label, MessageSenderEnum.HUMAN);
      } else {
        chatStore.appendMessage(
          [
            {
              label: label || message,
              sender: MessageSenderEnum.HUMAN,
              type: MessageTypeEnum.PLAIN,
            },
          ],
          MessageSenderEnum.HUMAN,
        );
      }
    }
  };
}
