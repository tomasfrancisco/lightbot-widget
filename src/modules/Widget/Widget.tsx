import throttle from "lodash.throttle";
import { inject, observer } from "mobx-react";
import DevTools from "mobx-react-devtools";
import * as React from "react";
import Frame from "react-frame-component";
import ReactResizeDetector from "react-resize-detector";
import styled from "styled-components";

import { highestZIndex } from "../../constants/highestZIndex";
import { Hotspot } from "../Hotspot/Hotspot";
import { FrameProvider } from "./FrameProvider";
import { ChatStore } from "./modules/ChatModal/stores/ChatStore";
import { Window } from "./Window";

type FrameWrapperProps = {
  isOpen: boolean;
  isMobileVersion: boolean;
};

const FrameWrapper = styled<FrameWrapperProps, "div">("div")`
  > iframe {
    margin: 0;
    padding: 0;
    position: fixed;
    border: 1px solid rgba(0, 0, 0, 0.12);
    border-radius: 8px;
    width: 400px;
    height: 0px;
    max-height: calc(100vh - 110px - 20px);
    bottom: 110px;
    right: 20px;
    box-shadow: 0 5px 40px rgba(0, 0, 0, 0.16);
    z-index: -1;
    opacity: 0;
    transition: all 0.5s ease-in-out;

    ${props =>
      props.isMobileVersion &&
      `
      bottom: 100;
      right: 0;
      left: 0;
      height: 0;
      width: 100vw;
      max-height: 100vh;
      border-radius: 0px;
      border-width: 0px;
    `};

    ${props =>
      props.isOpen &&
      `
      z-index: ${highestZIndex};
      height: calc(100vh - 35px);
      opacity: 1;
    `} ${props =>
      props.isOpen &&
      props.isMobileVersion &&
      `
      height: 100%;
      bottom: 0;
    `};
  }
`;

@inject("chatStore")
@observer
export class Widget extends React.Component<{ chatStore?: ChatStore }> {
  constructor(props) {
    super(props);

    this.onResizeHandler = throttle(this.onResizeHandler, 500);
  }

  public render() {
    const { chatStore } = this.props;
    if (!chatStore) {
      return null;
    }

    return (
      <>
        <Hotspot />
        {this.renderFrame()}
        {process.env.NODE_ENV === "development" && <DevTools />}
      </>
    );
  }

  private renderFrame() {
    const { chatStore } = this.props;
    if (!chatStore) {
      return null;
    }
    const { agentData$, isOpen$, isMobileVersion$ } = chatStore;

    return (
      <>
        <ReactResizeDetector handleWidth={true} onResize={this.onResizeHandler} />
        <FrameWrapper isOpen={isOpen$} isMobileVersion={isMobileVersion$}>
          <Frame>
            <FrameProvider>
              <Window agentId={agentData$.id} />
            </FrameProvider>
          </Frame>
        </FrameWrapper>
      </>
    );
  }

  private onResizeHandler = width => {
    const { chatStore } = this.props;
    if (chatStore) {
      chatStore.isMobileVersion$ = width < 420;
    }
  };
}
