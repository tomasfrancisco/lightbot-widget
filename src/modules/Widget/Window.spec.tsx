import { shallow } from "enzyme";
import { ChatStore } from "modules/Widget/modules/ChatModal";
import { Window } from "modules/Widget/Window";
import * as React from "react";
import { getBrowserStorageMock } from "utils/tests/browserStorageMock";
import { storeProvider } from "utils/tests/storeProvider";

describe("modules/Widget/Window", () => {
  let component;

  beforeEach(() => {
    component = shallow(
      storeProvider({
        chatStore: new ChatStore({ id: "agentId" }, getBrowserStorageMock()),
      })(<Window agentId={"agentId"} />),
    );
  });

  it("renders", () => {
    expect(component).toMatchSnapshot();
  });
});
