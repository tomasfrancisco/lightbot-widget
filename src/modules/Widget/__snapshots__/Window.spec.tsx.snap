// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/Widget/Window renders 1`] = `
<inject-Window-with-chatStore
  agentId="agentId"
/>
`;
